### PR TITLE
New cleanup

### PR DESCRIPTION
**Pull request by jurf**
This time in the suggested way.

Obsoletes #8.

Note that pyqt5 can be built with only selecting some modules, but I haven’t yet been able to figure which Anki needs (it didn’t want to run in my testing). @bochecha, ping. :slightly_smiling_face:

### CODE REVIEW

**Comment by jurf** [comment]
bot, build

**Comment by bochecha** [comment]
Hi @jurf, I'm trying this, it fails: the json points to anki.svg, but the file is named net.ankiweb.Anki.svg.

And then you have the `rename-icon` which is unneeded (and fails the build) as the icon is already properly named.

At the end of the build, there's this:

```
xdg-mime install anki.xml --novendor
make: xdg-mime: Command not found
make: [Makefile:19: install] Error 127 (ignored)
xdg-mime default anki.desktop application/x-anki
make: xdg-mime: Command not found
make: [Makefile:20: install] Error 127 (ignored)
xdg-mime default anki.desktop application/x-apkg
make: xdg-mime: Command not found
make: [Makefile:21: install] Error 127 (ignored)
```

**Comment by jurf** [comment]
@bochecha ah shoot, beginner mistake, but was too lazy to wait half an hour for it to build again at this hour. Lemme fix it quickly.

BTW the rename icon is there so that we don’t have to `sed` the desktop file. The less `sed`, the better. :D

**Comment by jurf** [comment]
OK, should be fixed, though I’m compiling now to be sure.

BTW, these are the modules that are built when the whole of PyQt is built:
```
QtCore, QtGui, QtHelp, QtMultimedia,
QtMultimediaWidgets, QtNetwork, QtOpenGL, QtPrintSupport, QtQml, QtQuick,
QtSql, QtSvg, QtTest, QtWidgets, QtXml, QtXmlPatterns, QtDesigner, QtDBus,
_QOpenGLFunctions_2_0, _QOpenGLFunctions_2_1, _QOpenGLFunctions_4_1_Core,
QtSensors, QtSerialPort, QtX11Extras, QtBluetooth, QtPositioning,
QtQuickWidgets, QtWebSockets, QtWebChannel, QtWebEngineWidgets, QtLocation,
QtNfc, QtWebEngineCore, QtWebEngine.
```

**Comment by bochecha** [comment]
You need to use `--no-deps` for all the pip installs.

As it is, you get things like this:

```
========================================================================
Building module python3-requests in /home/mathieu/.cache/flatpak-builder/build/python3-requests-1
========================================================================
Running: pip3 install --no-index --find-links="file://${PWD}" --prefix=${FLATPAK_DEST} requests
Ignoring indexes: https://pypi.python.org/simple
Collecting requests
Collecting chardet<3.1.0,>=3.0.2 (from requests)
Collecting idna<2.8,>=2.5 (from requests)
Collecting urllib3<1.24,>=1.21.1 (from requests)
Collecting certifi>=2017.4.17 (from requests)
Installing collected packages: chardet, idna, urllib3, certifi, requests
  Running setup.py install for chardet ... done
  Running setup.py install for idna ... done
  Running setup.py install for urllib3 ... done
  Running setup.py install for certifi ... done
  Running setup.py install for requests ... done
Successfully installed certifi-2018.8.24 chardet-3.0.4 idna-2.7 requests-2.19.1 urllib3-1.23
```

That means when you install the requests module, pip goes to PyPI and downloads the dependencies.

That's going to fail in Flathub, where builds can't access the network.

**Comment by jurf** [comment]
@bochecha pretty sure that should work, I used this for that as always: https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

There’s basically no difference to the updated version and the old versions (except for the newer packages), and those worked without issue.

**Comment by jurf** [comment]
bot, build

**Comment by bochecha** [comment]
As for PyQt modules, you need these:

```
            "  --enable=QtCore \\",
            "  --enable=QtGui \\",
            "  --enable=QtNetwork \\",
            "  --enable=QtPrintSupport \\",
            "  --enable=QtWebChannel \\",
            "  --enable=QtWebEngineCore \\",
            "  --enable=QtWebEngineWidgets \\",
            "  --enable=QtWidgets \\",
```

And then the app works. :slightly_smiling_face:

**Comment by jurf** [comment]
@bochecha thanks so much! I couldn’t figure that out for the life of me.

Regarding the `xdg-mime` errors, I’m not sure why they fail and why the mime file is installed by hand afterwards, I’ll have to look into that, but it doesn’t fail the build and it went through the initial review so I guess it’s OK…

File: python3-bs4.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    ],
+    "sources": [
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/88/df/86bffad6309f74f3ff85ea69344a078fc30003270c8df6894fca7a3c72ff/beautifulsoup4-4.6.3.tar.gz",
+            "sha256": "90f8e61121d6ae58362ce3bed8cd997efb00c914eae0ff3d363c32f9a9822d10"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/10/ed/7e8b97591f6f456174139ec089c769f89a94a1a4025fe967691de971f314/bs4-0.0.1.tar.gz",
```
Comment (bochecha) [inline python3-bs4.json]: Just remove this.

The module you want to install is beautifulsoup4, bs4 is just a dummy package to avoid name squatting.

(that was part of an interesting experiment years ago, where someone squatted easy typos on PyPI, Rubygems, NPM, … with packages which would send info to their servers about the IP of the machine installing their packages, the uid running the install, etc… very interesting study in how all those package managers are vulnerable :slightly_smiling_face:)

Comment (jurf) [reply python3-bs4.json]: Will do. :slightly_smiling_face:

**Comment by jurf** [comment]
bot, build

**Comment by dsd** [comment]
pyqt5 depends on sip and won't build without it
mpv depends on ffmpeg and won't build without it

That's why they are currently nested modules (also the mpv flatpak manifest is set up like this - https://github.com/flathub/io.github.GnomeMpv/blob/master/io.github.GnomeMpv.json). So I think they must stay nested, in order to represent the build dependencies. Please let me know if I'm missing something.

**Comment by jurf** [comment]
@bochecha nice, build times went down 22→14 minutes after that PyQt5 change. Thanks again!

@dsd manifests are processed in order, so that each module has access to whatever the previous module installed. E.g., if Anki actually compiled anything, it would fail with the manifest currently in master, because since it is at the start, _none_ of the dependencies are there yet, and it has nothing to compile against. They all install into the same root (`/app`).

Nested modules are an option, yes, but I don’t like using them, because they break the order, e.g. FFmpeg gets built before MPV, even though it’s _after_ it.

Nested modules also aren’t very popular in my experience (i.e. quite rare) on Flathub. If you sometimes wonder what’s the ‘correct’ way to do things, take a look at manifests done by (and maintained) the Flathub maintainers, e.g. TingPing, barthalion, etc.

**Comment by jurf** [comment]
But, in the end, it’s a matter of personal preference.

File: python3-bs4.json (hunk 1 above)
Comment (jurf) [reply python3-bs4.json]: dae/anki#257

File: net.ankiweb.Anki.json [hunk 2]
```diff
@@ -83,104 +125,66 @@
...
-          "cleanup": [
-            "/bin",
-            "/include"
-          ]
+          "type": "archive",
+          "url": "http://www.portaudio.com/archives/pa_stable_v190600_20161030.tgz",
+          "sha256": "f5a21d7dcd6ee84397446fa1fa1a0675bb2e8a4a6dceb4305a8404698d8d1513"
         }
+      ],
+      "cleanup": [
```
Comment (dsd) [inline net.ankiweb.Anki.json]: This portaudio cleanup is different from the original, so the output from the build differs in that respect. any chance you could limit this commit to purely stylistic changes, and move the cleanups to separate commits?

Comment (jurf) [reply net.ankiweb.Anki.json]: I don’t think this should change the build, just a simplification of the cleanup. Well, it won’t change the build, if the previous cleanup actually worked. `/app` is added automatically, so I’m not sure if this would double it or not, I didn’t think of that. I’ll check.

Comment (jurf) [reply net.ankiweb.Anki.json]: Yeah, the previous cleanup was broken, it didn’t clean anything, so this _does_ actually change the build.

Still, the change is purely cosmetic (along with a megabyte shaved off the Flatpak), it doesn’t affect the app function in any way, I’d just leave it TBH. But if you insist, I’ll try to summon my git-rebase-fu.

**Comment by jurf** [comment]
Ping?

File: net.ankiweb.Anki.json (hunk 2 above)
Comment (jurf) [reply net.ankiweb.Anki.json]: Summoned.

**Comment by jurf** [comment]
Hey what’s going on, why is it failing on ARM all of a sudden?
